core/lib/spree/migration_helpers.rb

```diff
@@ -0,0 +1,64 @@
+module Spree
+  module MigrationHelpers
+    class SafeIndexChange
+      attr_reader :migration, :table, :columns, :options
+
+      def initialize(migration, table, columns, options = {})
+        @migration = migration
+        @table = table
+        @columns = columns
+        @options = options
+      end
+
+      def remove_index
+        migration.remove_index(table, columns) if index_present?
+      end
+
+      def add_index
+        if columns.is_a?(Array)
+          add_multi_column_index
+        else
+          add_single_column_index
+        end
+      end
+
+      private
+
+      def index_present?
+        migration.index_exists?(table, columns, options)
+      end
+
+      def index_absent?
+        !index_present?
+      end
+
+      def add_multi_column_index
+        if all_columns_exist? && index_absent?
+          migration.add_index(table, columns, options)
+        end
+      end
+
+      def all_columns_exist?
+        columns.all? { |column| migration.column_exists?(table, column) }
```

Comment:
I agree with @cbrunsdon, this is a little too heavy. I would prefer the simpler `Array.wrap`, although I would probably just create a private method (instead of in-lining it) and call it a day:

```ruby
def safe_add_index(table, columns, options = {})
  if columns_exist?(table, columns) && !index_exists?(table, columns, options)
    add_index(table, columns, options)
  end
end

private

def columns_exist?(table, columns)
  Array.wrap(columns).all? { |column| column_exists?(table, column) }
end
```

I also don't think including the options in the `column_exits?` call is correct. The options passed in to that will check to see if there's a column with those restrictions. Whereas the options we're passing in are meant to apply to the index associated with that column, not the column itself.